File: rust/crates/numbers-utils/tests/operations/fibonacci_utils_tests.rs
use chrono::Local;
use colored::Colorize;

use num_bigint::BigUint;
use numbers_utils::operations::fibonacci_utils::{fibonacci, fibonacci_random};

/// # `format_error_message`
///
/// This function formats the error message for the test cases.
///
/// __Arguments:__
///
/// * `expected: BigUint` - The expected output.
/// * `actual: BigUint` - The actual output.
///
/// __Return:__
///
/// * `colored::ColoredString` - The formatted error message.
fn format_error_message(expected: &BigUint, actual: &BigUint) -> colored::ColoredString {
    format!(
        "\n\nTest run at: {}\nExpected : {}\nActual   : {}\n",
        Local::now().to_rfc3339(),
        expected,
        actual
    )
    .red()
}

/// # `fibonacci_test`
///
/// This function tests the `fibonacci` function.
#[test]
fn fibonacci_test() {
    let num = 100;
    let expected_output = BigUint::parse_bytes(b"354224848179261915075", 10).unwrap();
    let actual_output = fibonacci(num);
    let error_message = format_error_message(&expected_output, &actual_output);
    assert_eq!(expected_output, actual_output, "{}", error_message);
}

/// # `fibonacci_random_test`
///
/// This function tests the `fibonacci_random` function.
#[test]
fn fibonacci_random_test() {
    let result = fibonacci_random();
    let start_string = "Fib[";
    let error_message = format!("\nExpected: {}\n     Got: {}\n", start_string, result).red();
    assert!(result.starts_with(start_string), "{}", error_message,);
    assert!(result.contains(" = "));
}

File: haskell/packages/numbers-utils/tests/Operations/FibonacciUtilsTests.hs
{-# LANGUAGE OverloadedStrings #-}

module Operations.FibonacciUtilsTests (fibonacciTest, fibonacciRandomTest) where

import Data.List (isInfixOf, isPrefixOf)

import Data.Time.Clock (getCurrentTime)
import Data.Time.LocalTime (utcToLocalZonedTime)
import Test.Tasty (TestTree)
import Test.Tasty.HUnit (assertBool, assertEqual, testCase)

import Numbers.Utils.Operations.FibonacciUtils (fibonacci, fibonacciRandom)

-- |
-- # @formatErrorMessage@
--
-- The function @formatErrorMessage@ is used to format the error message.
--
-- __Arguments:__
--
-- * @expected: String@ - The expected output.
-- * @actual: String@ - The actual output.
--
-- __Return:__
--
-- * @IO String@ - The formatted error message.
formatErrorMessage :: String -> String -> IO String
formatErrorMessage expected actual = do
    now <- getCurrentTime >>= utcToLocalZonedTime
    return $
        "\nTest run at: "
            ++ show now
            ++ "\nExpected : "
            ++ expected
            ++ "\nActual   : "
            ++ actual
            ++ "\n"

-- |
-- # @fibonacciTest@
--
-- The function @fibonacciTest@ is used to test the @fibonacci@ function.
fibonacciTest :: TestTree
fibonacciTest = testCase "Fibonacci" $ do
    let num = 100
    let expectedOutput = read "354224848179261915075" :: Integer
    let actualOutput = fibonacci num
    errorMessage <- formatErrorMessage (show expectedOutput) (show actualOutput)
    assertEqual errorMessage expectedOutput actualOutput

-- |
-- # @fibonacciRandomTest@
--
-- The function @fibonacciRandomTest@ is used to test the @fibonacciRandom@
fibonacciRandomTest :: TestTree
fibonacciRandomTest = testCase "Fibonacci Random Test" $ do
    result <- fibonacciRandom
    let startString = "Fib["
    let errorMessage = "\nExpected: " ++ startString ++ "\n     Got: " ++ result ++ "\n"
    assertBool errorMessage (startString `isPrefixOf` result)
    assertBool errorMessage (" = " `isInfixOf` result)
